teps/0139-trusted-artifacts.md

```diff
@@ -0,0 +1,478 @@
+- This proposal does not discuss how to expose artifacts outside of a pipeline, even though it sets foundations that could be used to achieve that
+- This proposal does not discuss how to inject artifact as inputs to a pipeline, even though it sets foundations that could be used to achieve that. For instance, one could use a workspace preprovisioned with artifacts and use artifact type params as inputs for a pipeline
+
+### Use Cases
+
+- Extend the chain of trust across Tasks for provenance produced by Tekton Chains based on the `TaskRuns` and `PipelineRuns` executed by Tekton Pipeline
+
+### Requirements
+
+Support the following combinations:
+
+- One producer, one consumer
+- One producer, N consumers, including with write-one, read-many storage class
+- Many producers, one consumer
+- Many producers, many consumers, including with write-one, read-many storage class
+- Fail validation if workspaces (static/runtime) are not fit for artifacts
+- Fail execution if checksum validation fail, surface error to TaskRun/PipelineRun failure reason
+
+## Proposal
+
+In short the proposal is to:
+
+- extend parameters and result types to a new type `artifact`, an object type with a fixed schema
+- extend the `Pod` logic in the controller to inject hashing and checking steps when required
+- extend variable expansion to include `.data.path` for artifact parameters and results
+
+The same pipeline from demo, rewritten after implementation, is shown by this [demo pipeline](https://gist.github.com/afrittoli/7236be5fca524b752c221d2346497bb7):
+
+<details>
+  <summary>Demo Pipeline:</summary>
+
+```yaml=
+apiVersion: tekton.dev/v1
+kind: PipelineRun
+metadata:
+  generateName: trusted-artifacts-sugar
+spec:
+  pipelineSpec:
+    workspaces:
+      - name: artifactStorage  # In this example this is where we store artifacts
+        artifacts: true # this will result in failed validation if the workspace is bound to a readonly backend like a secret
+    tasks:
+      - name: producer
+        taskSpec:
+          results:
+            - name: aFileArtifact
+              type: artifact # inbuilt object schema (path, hash, type)
+              description: An artifact file
+            - name: aFolderArtifact
+              type: artifact # inbuilt object schema (path, hash, type)
+              description: An artifact folder
+          steps:
+            - name: produce-file
+              image: bash:latest
+              script: |
+                #!/usr/bin/env bash
+                # Produce some content. The result "data.path" will resolve to the workspace marked for artifacs.
+                # Since the workspace can only be mounted in write mode to one `Pod` at the time, it is safe
+                # to let the producer write directly into the workspace, thus saving data I/O
+                date +%s | tee "$(results.aFileArtifact.data.path)/afile.txt"
```

Comment:
Thinking out loud here. If the user is working in a different workspace then once their artifact is produced, they have to copy it to `$(results.aFileArtifact.data.path)` in order for it to be hashed.

Would it not be better if they could somehow indicate the absolute path of the artifact to the injected step. The injected step would have access to that `path` and do the copying to the `artifact workspace` and then compute the hash.  

Also, how do we safeguard against overwriting artifacts?
For example, if a user's task A creates an artifact `aFile.txt` and then task C overwrites that artifact `aFile.txt`. task D then want the artifact from task A, that will not work. We should be able to preserve the state of the artifact as a function of the task, right? Is this a valid goal or a non-goal?

Reply:
Thanks @chitrangpatel - I was under the false impression that we could mount the volume with write access to only a single `Pod`, but that is not possible, at least not until `ReadWriteOncePod` is available and even then it won't be generally available.

To solve that, we need the producer to write to an `emptyDir` first. The injected step will calculate the hash, and then copy the data to the workspace.  We cannot stop other tasks from corrupting the data but if that happens it will be reflected by an invalid hash.

Reply:
One last follow up (if I'm understanding this correctly):
If an artifact is overwritten, it cannot be requested from its previous state. Otherwise, it will result in an error due to a different `hash`. May be this is something we should highlight in the doc. 

For example, if `task1` produced an artifactFile. It was consumed by `task2` which modified the artifact. Now, if `task3` wants the artifact, it can only correctly get it by requesting the artifact using the hash outputted by `task2`. If it requests the hash of the artifact using `task1`'s result then it will result in a failure.